move nbt copying to mixin

KyllianGamer opened the pull request:
Instead of using a tick event, the copying now uses a key pressed event. This does what #22 requests. The only difference being it's not with any rendering code, as we had discussed that in discord.

KyllianGamer committed:
move nbt copying to mixin

Instead of using a tick event, the copying now uses a key pressed event.

KyllianGamer commented:
I just want to add that it also wouldn't be hard to either change the static binding or make an actual keybind for it. Let me know what you guys think.

Also, this is just a pure translation from the tick event. This shouldn't be the final version. Things that could/should be changed:
- Non-static binding or different binding.
- Indication of having copied the NBT outside of the logging in the console. Right now when you copy the NBT you can't be sure if it actually copied it until you tested it. Perhaps some chat message could fix this.
- Different formatting. Right now the formatting is "minecraft: [ITEM NAME] {NBT}". If you want to do the "minecraft:" thing you need to add the item's translation key and not its name. 
- Potential bug fixes.

KyllianGamer committed:
item path

Changed the copied text to use the identifier of the item instead of its unformatted name.

Commit:
Slight update to how the NBT copying is shown to the user with visible feedback.

wiviw5 commented:
I added some visible feedback for the activation to show the user they copied something to the clipboard.

KyllianGamer committed:
item lore hover text

Adds the lore of the item as hover text on the item's name in the nbt copy text, because why not.

KyllianGamer commented:
Ready for review

TGWaffles committed:
add comments, javadoc, and remove unnecessary for-loop

TGWaffles commented on src/main/java/io/github/skyblockcore/mixin/HandledScreenMixin.java:
```diff
@@ -0,0 +1,55 @@
+        if (MinecraftClient.getInstance().player == null) return;
+        MutableText tooltip = Text.literal("");
+        List<Text> loreLines = itemToCopyNBT.getTooltip(MinecraftClient.getInstance().player, TooltipContext.BASIC);
+        for (int i = 0; i<loreLines.size(); i++)
+            tooltip.append((i==0) ? "" : "\n").append(loreLines.get(i));
```
This doesn't seem necessary. Just display the item on hover.

TGWaffles commented on src/main/java/io/github/skyblockcore/mixin/HandledScreenMixin.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Shadow @Final protected ScreenHandler handler;
+
+    @Inject(method = "keyPressed", at = @At("HEAD"))
+    void onKeyPressed(int keyCode, int scanCode, int modifiers, CallbackInfoReturnable<Boolean> cir) {
+        if (!SkyblockCoreCommand.NBTCOPYING) return;
+        if (GLFW.GLFW_KEY_RIGHT_CONTROL != keyCode) return;
+        Slot focused = this.focusedSlot;
+        if (focused == null || !focused.hasStack() || !handler.getCursorStack().isEmpty()) return;
+        ItemStack itemToCopyNBT = focusedSlot.getStack();
+        if (itemToCopyNBT.getNbt() == null) return;
+        String itemNBT = "minecraft:" + itemToCopyNBT.getItem().toString() + " " + itemToCopyNBT.getNbt();
+        if (ConfigManager.getConfig() != null && ConfigManager.getConfig().isDev()) SkyblockCore.LOGGER.info(TITLE + " [Dev NBT] > " + itemNBT);
+        MinecraftClient.getInstance().keyboard.setClipboard(itemNBT);
+        if (MinecraftClient.getInstance().player == null) return;
+        MutableText tooltip = Text.literal("");
+        List<Text> loreLines = itemToCopyNBT.getTooltip(MinecraftClient.getInstance().player, TooltipContext.BASIC);
+        for (int i = 0; i<loreLines.size(); i++)
+            tooltip.append((i==0) ? "" : "\n").append(loreLines.get(i));
+        MinecraftClient.getInstance().player.sendMessage(Text.literal(TITLE + " NBT Data has been copied for: ").formatted(Formatting.WHITE).append(((MutableText) itemToCopyNBT.getName()).styled(s -> s.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, tooltip)))), false);
+    }
```
Some comments for this would be nice.

TGWaffles commented on src/main/java/io/github/skyblockcore/mixin/HandledScreenMixin.java:
```diff
@@ -0,0 +1,78 @@
+        // Ensure the NBT copying functionality is enabled.
+        if (!SkyblockCoreCommand.NBTCOPYING) return;
+
+        // Check that the player has pressed the right control key (this should be editable in the future).
+        if (GLFW.GLFW_KEY_RIGHT_CONTROL != keyCode) return;
```
Is it worth making this key customisable before we merge this?

Ascynx replied:
Should be imo, would stop possible conflicts

KyllianGamer replied:
This has now been added.

KyllianGamer committed:
binding added and no nbt warning

Added a binding for the copying of the nbt and also added the sending of a message when no NBT is present on an item. This is so people won't think the feature doesn't work when there is no nbt present. (This totally didn't happen to me)

KyllianGamer committed:
remove nbt "tip"

Removed the part of the message giving people a tip of how to copy the nbt by telling them the key. This is no longer needed due to having added a binding.
